src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -2241,7 +2636,85 @@ private function step_in_row(): bool {
 	 * @return bool Whether an element was found.
 	 */
 	private function step_in_cell(): bool {
-		$this->bail( 'No support for parsing in the ' . WP_HTML_Processor_State::INSERTION_MODE_IN_CELL . ' state.' );
+		$tag_name = $this->get_tag();
+		$op_sigil = $this->is_tag_closer() ? '-' : '+';
+		$op       = "{$op_sigil}{$tag_name}";
+
+		switch ( $op ) {
+			/*
+			 * > An end tag whose tag name is one of: "td", "th"
+			 */
+			case '-TD':
+			case '-TH':
+				if ( ! $this->state->stack_of_open_elements->has_element_in_table_scope( $tag_name ) ) {
+					// @todo Indicate a parse error once it's possible.
+					return $this->step();
+				}
+				$this->generate_implied_end_tags();
+				if ( ! $this->state->stack_of_open_elements->current_node()->node_name ) {
+					// @todo Indicate a parse error once it's possible.
+				}
+				$this->state->stack_of_open_elements->pop_until( $tag_name );
+				$this->state->active_formatting_elements->clear_up_to_last_marker();
+				$this->state->insertion_mode = WP_HTML_Processor_State::INSERTION_MODE_IN_ROW;
+				return true;
+
+			/*
+			 * > A start tag whose tag name is one of: "caption", "col", "colgroup", "tbody", "td",
+			 * > "tfoot", "th", "thead", "tr"
+			 */
+			case '+CAPTION':
+			case '+COL':
+			case '+COLGROUP':
+			case '+TBODY':
+			case '+TD':
+			case '+TFOOT':
+			case '+TH':
+			case '+THEAD':
+			case '+TR':
+				// Assert: The stack of open elements has a td or th element in table scope.
+				if (
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TD' ) &&
+					! $this->state->stack_of_open_elements->has_element_in_table_scope( 'TH' )
+				) {
+					throw new Exception( 'Assertion failed @todo better message' );
```

Comment:
earmark: we shouldn't leave the exception in here

Reply:
This is an assertion from the specification. Should we bail? Should we have an assert method?

If this assertion fails, it means something has gone very wrong and it's not clear what we can do other than fail.

Reply:
tell me what you think.
of course something has gone wrong, but I also don't want to throw an `Exception()`

how likely is it that we'd get here? is it dependent on the document, or on our code?
what are the repercussions of going forward past this assertion if it fails?

Reply:
thanks for gathering all this. I think if we want to assert this we should do so from a test or through running tests to parse documents in the wild. should we encounter this situation on a production site, there's apparently nothing anyone can do about it, so I think closing too many elements is not the worst failover for a condition that should never trigger